File: windows/SpeechX/Core/DeepgramService.cs
using System.Diagnostics;
using System.Net.Http;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SpeechX.Core;

/// <summary>
/// Streams 16 kHz mono linear-PCM16 audio to Deepgram over a WebSocket and surfaces interim +
/// final transcripts. Port of the macOS DeepgramService (URLSessionWebSocketTask -> ClientWebSocket).
/// </summary>
public sealed class DeepgramService
{
    private const string ListenUrl = "wss://api.deepgram.com/v1/listen";
    private const string ModelsUrl = "https://api.deepgram.com/v1/models";

    /// <summary>Cap pre-connection buffer at ~10s of 16 kHz mono Int16 audio.</summary>
    private const int MaxBufferedBytes = 16_000 * 2 * 10;

    /// <summary>
    /// Deepgram caps keyterm prompting at 500 tokens per request; keep well under that and avoid
    /// an oversized query string by sending at most this many terms.
    /// </summary>
    private const int MaxKeyterms = 100;

    private static readonly HttpClient Http = new();

    /// <summary>
    /// Fired whenever the live transcript changes (interim or final). The argument is the running
    /// transcript from the start of the current session including the latest interim guess.
    /// May fire on a background thread — the consumer marshals to the UI thread.
    /// </summary>
    public Action<string>? OnPartialTranscript;

    private readonly object _lock = new();
    private ClientWebSocket? _ws;
    private CancellationTokenSource? _cts;
    private readonly SemaphoreSlim _sendLock = new(1, 1);

    private string _accumulatedTranscript = "";
    private string _currentInterim = "";
    private bool _isWaitingForFinal;
    private TaskCompletionSource<string>? _finalTcs;
    private CancellationTokenSource? _timeoutCts;

    private readonly List<byte[]> _pendingFrames = new();
    private bool _isSocketOpen;

    public void Connect(string apiKey, string model, string language, IReadOnlyList<string>? keyterms = null)
    {
        lock (_lock)
        {
            _accumulatedTranscript = "";
            _currentInterim = "";
            _isWaitingForFinal = false;
            _finalTcs = null;
            _pendingFrames.Clear();
            _isSocketOpen = false;
        }
        EmitPartial();

        if (string.IsNullOrEmpty(apiKey)) return;

        var queryParts = new List<string>
        {
            "encoding=linear16",
            "sample_rate=16000",
            "channels=1",
            $"model={Uri.EscapeDataString(model)}",
            $"language={Uri.EscapeDataString(language)}",
            "punctuate=true",
            "interim_results=true",
        };

        // Keyterm prompting biases recognition toward the user's focus words. Gate strictly on
        // Nova-3 + English (our default, and the config we've confirmed accepts the param): this
        // query rides the same handshake all dictation depends on, so an unsupported param would
        // 400 and break every recording. The LLM glossary still corrects spelling for every other
        // model/language, so a tight gate costs only the recognition boost, never the feature.
        if (model.StartsWith("nova-3") && language.StartsWith("en") && keyterms != null)
        {
            foreach (var term in keyterms.Take(MaxKeyterms))
            {
                var trimmed = term.Trim();
                if (trimmed.Length == 0) continue;
                queryParts.Add($"keyterm={Uri.EscapeDataString(trimmed)}");
            }
        }

        var uri = new UriBuilder(ListenUrl)
        {
            Query = string.Join('&', queryParts),
        }.Uri;

        var ws = new ClientWebSocket();
        ws.Options.SetRequestHeader("Authorization", $"Token {apiKey}");
        _ws = ws;
        _cts = new CancellationTokenSource();

        _ = RunAsync(ws, uri, _cts.Token);
    }

    private async Task RunAsync(ClientWebSocket ws, Uri uri, CancellationToken ct)
    {
        try
        {
            await ws.ConnectAsync(uri, ct).ConfigureAwait(false);
        }
        catch (Exception e)
        {
            Debug.WriteLine($"[deepgram] connect failed: {e.Message}");
            DeliverAndDisconnect();
            return;
        }

        // Flush any frames captured during the handshake, in order.
        byte[][] frames;
        lock (_lock)
        {
            if (!ReferenceEquals(ws, _ws)) return; // superseded by a newer connection
            frames = _pendingFrames.ToArray();
            _pendingFrames.Clear();
            _isSocketOpen = true;
        }
        foreach (var frame in frames)
            await SendFrameAsync(frame, ct).ConfigureAwait(false);

        await ReceiveLoopAsync(ws, ct).ConfigureAwait(false);
    }

    /// <summary>Send one PCM16 audio frame. Frames captured before the socket opens are buffered.</summary>
    public void SendAudio(byte[] data)
    {
        bool openNow;
        lock (_lock)
        {
            if (_ws == null) return;
            openNow = _isSocketOpen;
            if (!openNow)
            {
                int total = _pendingFrames.Sum(f => f.Length);
                while (total + data.Length > MaxBufferedBytes && _pendingFrames.Count > 0)
                {
                    total -= _pendingFrames[0].Length;
                    _pendingFrames.RemoveAt(0);
                }
                _pendingFrames.Add(data);
            }
        }

        if (openNow)
            _ = SendFrameAsync(data, _cts?.Token ?? CancellationToken.None);
    }

    private async Task SendFrameAsync(byte[] data, CancellationToken ct)
    {
        var ws = _ws;
        if (ws == null || ws.State != WebSocketState.Open) return;
        try
        {
            await _sendLock.WaitAsync(ct).ConfigureAwait(false);
            try
            {
                await ws.SendAsync(new ArraySegment<byte>(data), WebSocketMessageType.Binary, true, ct)
                        .ConfigureAwait(false);
            }
            finally { _sendLock.Release(); }
        }
        catch (Exception e)
        {
            Debug.WriteLine($"[deepgram] send failed: {e.Message}");
        }
    }

    public Task<string> CloseStreamAsync()
    {
        var tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
        lock (_lock)
        {
            _finalTcs = tcs;
            _isWaitingForFinal = true;
        }

        // Empty binary frame = Deepgram's signal to flush and finalize.
        _ = SendFrameAsync(Array.Empty<byte>(), _cts?.Token ?? CancellationToken.None);

        // Safety timeout: deliver what we have after 3 seconds.
        _timeoutCts = new CancellationTokenSource();
        var timeoutToken = _timeoutCts.Token;
        _ = Task.Delay(3000, timeoutToken).ContinueWith(t =>
        {
            if (!t.IsCanceled) DeliverAndDisconnect();
        }, TaskScheduler.Default);

        return tcs.Task;
    }

    /// <summary>Tear down without waiting for a final transcript (e.g. mic capture failed to start).</summary>
    public void Cancel()
    {
        lock (_lock)
        {
            _timeoutCts?.Cancel();
            _timeoutCts = null;
            _finalTcs = null;
            _isWaitingForFinal = false;
            _accumulatedTranscript = "";
            _currentInterim = "";
            _pendingFrames.Clear();
            _isSocketOpen = false;
        }
        Disconnect();
    }

    private async Task ReceiveLoopAsync(ClientWebSocket ws, CancellationToken ct)
    {
        var buffer = new byte[16 * 1024];
        var sb = new StringBuilder();
        try
        {
            while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
            {
                WebSocketReceiveResult result;
                try
                {
                    result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), ct).ConfigureAwait(false);
                }
                catch (Exception e)
                {
                    Debug.WriteLine($"[deepgram] receive failed: {e.Message}");
                    if (_isWaitingForFinal) DeliverAndDisconnect();
                    return;
                }

                if (result.MessageType == WebSocketMessageType.Close)
                {
                    if (_isWaitingForFinal) DeliverAndDisconnect();
                    return;
                }

                sb.Append(Encoding.UTF8.GetString(buffer, 0, result.Count));
                if (result.EndOfMessage)
                {
                    HandleMessage(sb.ToString());
                    sb.Clear();
                }
            }
        }
        catch (OperationCanceledException) { /* expected on teardown */ }
    }

    private void HandleMessage(string json)
    {
        DeepgramResponse? response;
        try { response = JsonSerializer.Deserialize<DeepgramResponse>(json); }
        catch { return; }
        if (response == null) return;

        string transcript = response.Channel?.Alternatives is { Count: > 0 } alts
            ? alts[0].Transcript ?? ""
            : "";

        if (response.IsFinal == true)
        {
            if (!string.IsNullOrEmpty(transcript))
            {
                lock (_lock)
                {
                    if (_accumulatedTranscript.Length > 0) _accumulatedTranscript += " ";
                    _accumulatedTranscript += transcript;
                }
            }
            lock (_lock) { _currentInterim = ""; }
            EmitPartial();
        }
        else
        {
            lock (_lock) { _currentInterim = transcript; }
            EmitPartial();
        }

        if (_isWaitingForFinal && response.IsFinal == true && response.SpeechFinal == true)
        {
            _timeoutCts?.Cancel();
            DeliverAndDisconnect();
        }
    }

    private void EmitPartial()
    {
        string combined;
        lock (_lock)
        {
            if (_currentInterim.Length == 0) combined = _accumulatedTranscript;
            else if (_accumulatedTranscript.Length == 0) combined = _currentInterim;
            else combined = _accumulatedTranscript + " " + _currentInterim;
        }
        OnPartialTranscript?.Invoke(combined);
    }

    private void DeliverAndDisconnect()
    {
        TaskCompletionSource<string>? tcs;
        string transcript;
        lock (_lock)
        {
            if (!_isWaitingForFinal) return;
            _isWaitingForFinal = false;
            transcript = _accumulatedTranscript;
            tcs = _finalTcs;
            _finalTcs = null;
        }
        tcs?.TrySetResult(transcript);
        Disconnect();
    }

    private void Disconnect()
    {
        var ws = _ws;
        _ws = null;
        try { _cts?.Cancel(); } catch { }
        try { ws?.Abort(); } catch { }
        try { ws?.Dispose(); } catch { }
    }

    public async Task<IReadOnlyList<DeepgramModel>> FetchModelsAsync(string apiKey)
    {
        if (string.IsNullOrEmpty(apiKey)) throw ApiException.MissingKey();

        using var request = new HttpRequestMessage(HttpMethod.Get, ModelsUrl);
        request.Headers.TryAddWithoutValidation("Authorization", $"Token {apiKey}");

        HttpResponseMessage response;
        string body;
        try
        {
            response = await Http.SendAsync(request).ConfigureAwait(false);
            body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
        }
        catch (Exception e)
        {
            throw ApiException.Network(e.Message);
        }

        if (!response.IsSuccessStatusCode)
            throw ApiException.Http((int)response.StatusCode, body);

        ModelsResponse? root;
        try { root = JsonSerializer.Deserialize<ModelsResponse>(body); }
        catch { throw ApiException.Decoding(); }
        if (root == null) throw ApiException.Decoding();

        var streamingSupport = new Dictionary<string, bool>();
        var displayNames = new Dictionary<string, string>();
        var languageMap = new Dictionary<string, List<string>>();

        foreach (var m in root.Stt ?? new())
        {
            var canonical = m.CanonicalName;
            if (string.IsNullOrEmpty(canonical)) continue;
            streamingSupport[canonical] = (streamingSupport.GetValueOrDefault(canonical)) || (m.Streaming ?? false);
            if (!displayNames.ContainsKey(canonical)) displayNames[canonical] = m.Name ?? canonical;
            var existing = languageMap.GetValueOrDefault(canonical) ?? new();
            foreach (var lang in m.Languages ?? new())
                if (!existing.Contains(lang)) existing.Add(lang);
            languageMap[canonical] = existing;
        }

        var models = streamingSupport
            .Where(kv => kv.Value)
            .Select(kv =>
            {
                var canonical = kv.Key;
                var langs = (languageMap.GetValueOrDefault(canonical) ?? new()).OrderBy(x => x, StringComparer.Ordinal).ToList();
                if (canonical.StartsWith("nova-2") || canonical.StartsWith("nova-3"))
                    langs.Add("multi");
                return new DeepgramModel(canonical, displayNames.GetValueOrDefault(canonical) ?? canonical, langs);
            })
            .OrderBy(m => m.CanonicalName, StringComparer.Ordinal)
            .ToList();

        return models;
    }

    // MARK: - Response DTOs

    private sealed class DeepgramResponse
    {
        [JsonPropertyName("channel")] public DeepgramChannel? Channel { get; set; }
        [JsonPropertyName("is_final")] public bool? IsFinal { get; set; }
        [JsonPropertyName("speech_final")] public bool? SpeechFinal { get; set; }
    }

    private sealed class DeepgramChannel
    {
        [JsonPropertyName("alternatives")] public List<DeepgramAlternative>? Alternatives { get; set; }
    }

    private sealed class DeepgramAlternative
    {
        [JsonPropertyName("transcript")] public string? Transcript { get; set; }
        [JsonPropertyName("confidence")] public double? Confidence { get; set; }
    }

    private sealed class ModelsResponse
    {
        [JsonPropertyName("stt")] public List<ModelEntry>? Stt { get; set; }
        [JsonPropertyName("tts")] public List<ModelEntry>? Tts { get; set; }
    }

    private sealed class ModelEntry
    {
        [JsonPropertyName("name")] public string? Name { get; set; }
        [JsonPropertyName("canonical_name")] public string? CanonicalName { get; set; }
        [JsonPropertyName("streaming")] public bool? Streaming { get; set; }
        [JsonPropertyName("languages")] public List<string>? Languages { get; set; }
    }
}

File: ios/SpeechXKeyboard/DeepgramService.swift
import Foundation
import AVFoundation
import os.log

private let dgLogger = Logger(subsystem: "com.speechx.app", category: "deepgram")

// MARK: - Public model type

struct DeepgramModel: Identifiable, Hashable {
    let canonicalName: String
    let displayName: String
    let languages: [String]
    var id: String { canonicalName }
}

// MARK: - Service

class DeepgramService: NSObject, URLSessionWebSocketDelegate {
    private static let listenURL = URL(staticString: "wss://api.deepgram.com/v1/listen")
    private static let modelsURL = URL(staticString: "https://api.deepgram.com/v1/models")

    /// Fired whenever the live transcript changes (interim or final). Always called on the main queue.
    /// The string is the running transcript from the start of the current session, including the
    /// latest interim guess, so it's safe to bind directly to UI state.
    var onPartialTranscript: ((String) -> Void)?

    private var webSocketTask: URLSessionWebSocketTask?
    private var session: URLSession!
    private var accumulatedTranscript: String = ""
    private var currentInterim: String = ""
    private var finalTranscriptCallback: ((String) -> Void)?
    private var isWaitingForFinal = false
    private var timeoutWorkItem: DispatchWorkItem?

    // Audio frames captured before the WebSocket finishes its handshake are
    // buffered here and flushed in order once the socket opens. Without this
    // the first ~100-300ms of speech is silently dropped on every recording.
    private var pendingFrames: [Data] = []
    private var isSocketOpen = false
    private let bufferLock = NSLock()
    /// Cap pre-connection buffer at ~10s of 16-kHz mono Int16 audio so a stuck
    /// handshake can't grow it without bound.
    private static let maxBufferedBytes = 16_000 * 2 * 10

    override init() {
        super.init()
        session = URLSession(
            configuration: .default,
            delegate: self,
            delegateQueue: nil
        )
    }

    /// Deepgram caps keyterm prompting at 500 tokens per request; keep well under that and avoid
    /// an oversized query string by sending at most this many terms.
    private static let maxKeyterms = 100

    func connect(apiKey: String, model: String, language: String, keyterms: [String] = []) {
        accumulatedTranscript = ""
        currentInterim = ""
        isWaitingForFinal = false
        finalTranscriptCallback = nil
        bufferLock.lock()
        pendingFrames.removeAll()
        isSocketOpen = false
        bufferLock.unlock()
        emitPartial()

        guard !apiKey.isEmpty else { return }

        guard var components = URLComponents(url: Self.listenURL, resolvingAgainstBaseURL: false) else {
            preconditionFailure("Invalid Deepgram listen URL")
        }
        var queryItems = [
            URLQueryItem(name: "encoding",        value: "linear16"),
            URLQueryItem(name: "sample_rate",     value: "16000"),
            URLQueryItem(name: "channels",        value: "1"),
            URLQueryItem(name: "model",           value: model),
            URLQueryItem(name: "language",        value: language),
            URLQueryItem(name: "punctuate",       value: "true"),
            URLQueryItem(name: "interim_results", value: "true"),
        ]

        // Keyterm prompting biases recognition toward the user's focus words. Gate strictly on
        // Nova-3 + English (our default, and the config we've confirmed accepts the param): this
        // query rides the same handshake all dictation depends on, so an unsupported param would
        // 400 and break every recording. The LLM glossary still corrects spelling for every other
        // model/language, so a tight gate costs only the recognition boost, never the feature.
        if model.hasPrefix("nova-3") && language.hasPrefix("en") {
            for term in keyterms.prefix(Self.maxKeyterms) {
                let trimmed = term.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !trimmed.isEmpty else { continue }
                queryItems.append(URLQueryItem(name: "keyterm", value: trimmed))
            }
        }
        components.queryItems = queryItems

        guard let url = components.url else {
            preconditionFailure("Failed to build Deepgram listen URL from components")
        }
        var request = URLRequest(url: url)
        request.setValue("Token \(apiKey)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 30

        webSocketTask = session.webSocketTask(with: request)
        webSocketTask?.resume()
        dgLogger.info("WebSocket connecting (model=\(model), language=\(language))")
        receiveNext()
    }

    func fetchModels(apiKey: String) async throws -> [DeepgramModel] {
        guard !apiKey.isEmpty else { throw APIError.missingKey }

        var request = URLRequest(url: Self.modelsURL)
        request.setValue("Token \(apiKey)", forHTTPHeaderField: "Authorization")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            dgLogger.error("/v1/models network error: \(error.localizedDescription)")
            throw APIError.network(error.localizedDescription)
        }

        let httpStatus = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(httpStatus) else {
            let body = String(data: data, encoding: .utf8)
            dgLogger.error("/v1/models HTTP \(httpStatus): \(body ?? "<binary>")")
            throw APIError.http(status: httpStatus, body: body)
        }
        guard let root = try? JSONDecoder().decode(ModelsResponse.self, from: data) else {
            let body = String(data: data, encoding: .utf8)
            dgLogger.error("/v1/models decode failed: \(body ?? "<binary>")")
            throw APIError.decoding
        }

        // Group by canonical_name; collect languages union, streaming if any variant supports it
        var streamingSupport: [String: Bool] = [:]
        var displayNames: [String: String] = [:]
        var languageMap: [String: [String]] = [:]
        for m in root.stt ?? [] {
            guard let canonical = m.canonicalName, !canonical.isEmpty else { continue }
            streamingSupport[canonical] = (streamingSupport[canonical] ?? false) || (m.streaming ?? false)
            if displayNames[canonical] == nil { displayNames[canonical] = m.name }
            let existing = languageMap[canonical] ?? []
            let newLangs = (m.languages ?? []).filter { !existing.contains($0) }
            languageMap[canonical] = existing + newLangs
        }
        let models = streamingSupport
            .filter { $0.value }
            .map { (canonical, _) in
                var langs = (languageMap[canonical] ?? []).sorted()
                // Inject "multi" for nova-2 and nova-3 families (not returned by API)
                if canonical.hasPrefix("nova-2") || canonical.hasPrefix("nova-3") {
                    langs.append("multi")
                }
                return DeepgramModel(
                    canonicalName: canonical,
                    displayName: displayNames[canonical] ?? canonical,
                    languages: langs
                )
            }
            .sorted { $0.canonicalName < $1.canonicalName }
        dgLogger.info("/v1/models returned \(models.count) streaming models")
        return models
    }

    func sendAudioBuffer(_ buffer: AVAudioPCMBuffer, format: AVAudioFormat) {
        guard let task = webSocketTask else { return }
        guard let channelData = buffer.int16ChannelData else { return }

        let frameCount = Int(buffer.frameLength)
        let data = Data(bytes: channelData[0], count: frameCount * MemoryLayout<Int16>.size)

        bufferLock.lock()
        let openNow = isSocketOpen
        if !openNow {
            // Drop oldest frames if a stuck handshake would push us over the cap.
            var totalBytes = pendingFrames.reduce(0) { $0 + $1.count }
            while totalBytes + data.count > Self.maxBufferedBytes && !pendingFrames.isEmpty {
                totalBytes -= pendingFrames.removeFirst().count
            }
            pendingFrames.append(data)
        }
        bufferLock.unlock()

        if openNow {
            task.send(.data(data)) { _ in }
        }
    }

    func closeStream() async -> String {
        await withCheckedContinuation { (continuation: CheckedContinuation<String, Never>) in
            finalTranscriptCallback = { transcript in
                continuation.resume(returning: transcript)
            }
            isWaitingForFinal = true

            // Send empty binary frame — Deepgram's signal to flush and finalize
            webSocketTask?.send(.data(Data())) { [weak self] error in
                if error != nil {
                    self?.deliverAndDisconnect()
                }
            }

            // Safety timeout: deliver what we have after 3 seconds
            let workItem = DispatchWorkItem { [weak self] in
                guard let self, self.isWaitingForFinal else { return }
                self.deliverAndDisconnect()
            }
            timeoutWorkItem = workItem
            DispatchQueue.global().asyncAfter(deadline: .now() + 3.0, execute: workItem)
        }
    }

    /// Tear down a connection without waiting for any final transcript. Safe to call
    /// when an upstream failure (e.g. mic capture failed to start) means there's
    /// nothing to deliver.
    func cancel() {
        timeoutWorkItem?.cancel()
        timeoutWorkItem = nil
        finalTranscriptCallback = nil
        isWaitingForFinal = false
        accumulatedTranscript = ""
        currentInterim = ""
        bufferLock.lock()
        pendingFrames.removeAll()
        isSocketOpen = false
        bufferLock.unlock()
        disconnect()
    }

    private func receiveNext() {
        webSocketTask?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                if case .string(let text) = message {
                    self.handleMessage(text)
                }
                self.receiveNext()
            case .failure(let error):
                dgLogger.error("WebSocket receive failed: \(error.localizedDescription)")
                if self.isWaitingForFinal {
                    self.deliverAndDisconnect()
                }
            }
        }
    }

    private func handleMessage(_ json: String) {
        guard let data = json.data(using: .utf8),
              let response = try? JSONDecoder().decode(DeepgramResponse.self, from: data) else {
            return
        }

        let transcript = response.channel?.alternatives?.first?.transcript ?? ""

        if response.isFinal == true {
            if !transcript.isEmpty {
                if !accumulatedTranscript.isEmpty {
                    accumulatedTranscript += " "
                }
                accumulatedTranscript += transcript
            }
            // Final result supersedes any interim guess for this utterance.
            currentInterim = ""
            emitPartial()
        } else {
            // Interim: each message replaces the previous interim guess for the current utterance.
            currentInterim = transcript
            emitPartial()
        }

        // speech_final = true means Deepgram received our stream-close signal and is done
        if isWaitingForFinal && response.isFinal == true && response.speechFinal == true {
            timeoutWorkItem?.cancel()
            deliverAndDisconnect()
        }
    }

    private func emitPartial() {
        let combined: String
        if currentInterim.isEmpty {
            combined = accumulatedTranscript
        } else if accumulatedTranscript.isEmpty {
            combined = currentInterim
        } else {
            combined = accumulatedTranscript + " " + currentInterim
        }
        let callback = onPartialTranscript
        DispatchQueue.main.async { callback?(combined) }
    }

    private func deliverAndDisconnect() {
        guard isWaitingForFinal else { return }
        isWaitingForFinal = false
        let transcript = accumulatedTranscript
        let callback = finalTranscriptCallback
        finalTranscriptCallback = nil
        callback?(transcript)
        disconnect()
    }

    private func disconnect() {
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
    }
}

// MARK: - URLSessionWebSocketDelegate

extension DeepgramService {
    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocolName: String?
    ) {
        bufferLock.lock()
        // Guard against a stale delegate callback for a previous task that
        // opened after we already cancelled and reconnected.
        guard webSocketTask === self.webSocketTask else {
            bufferLock.unlock()
            return
        }
        let frames = pendingFrames
        pendingFrames.removeAll()
        isSocketOpen = true
        bufferLock.unlock()

        if !frames.isEmpty {
            dgLogger.info("WebSocket opened — flushing \(frames.count) buffered audio frames")
            for frame in frames {
                webSocketTask.send(.data(frame)) { _ in }
            }
        }
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        if isWaitingForFinal {
            deliverAndDisconnect()
        }
    }
}

// MARK: - Response Models

private struct DeepgramResponse: Codable {
    let channel: DeepgramChannel?
    let isFinal: Bool?
    let speechFinal: Bool?

    enum CodingKeys: String, CodingKey {
        case channel
        case isFinal     = "is_final"
        case speechFinal = "speech_final"
    }
}

private struct DeepgramChannel: Codable {
    let alternatives: [DeepgramAlternative]?
}

private struct DeepgramAlternative: Codable {
    let transcript: String?
    let confidence: Double?
}

private struct ModelsResponse: Codable {
    let stt: [ModelEntry]?
    let tts: [ModelEntry]?
}

private struct ModelEntry: Codable {
    let name: String?
    let canonicalName: String?
    let streaming: Bool?
    let languages: [String]?

    enum CodingKeys: String, CodingKey {
        case name
        case canonicalName = "canonical_name"
        case streaming
        case languages
    }
}
